### PR TITLE
[CP Auth 3/5]: add CP client and catalog service

### PR DESCRIPTION
**Pull request by axbarsan**
Part of the larger #1547

This PR adds the CP client library, client and handlers.
It also links the newly created types throughout the app.

### CODE REVIEW

File: src/model/clients/FetchAdapter.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import { GenericResponse } from 'model/clients/GenericResponse';
+import { HttpClient } from 'model/clients/HttpClient';
+
+/**
+ * This is an adapter for using an HttpClient with the same
+ * interface as the global `fetch` function.
+ */
+class FetchAdapter<T extends HttpClient = HttpClient> {
```
Comment (axbarsan) [inline src/model/clients/FetchAdapter.ts:8]: This makes us able to use our classes based on the `HttpClient` base class with the CP Client, as a replacement for the global `fetch` method.

File: src/model/clients/GenericResponse.js [hunk 2]
```diff
@@ -103,4 +103,26 @@ export class GenericResponse {
   get headers() {
     return Object.assign({}, this.config.headers);
   }
+
+  /**
+   * Convert the object to a `Response` object returned by the fetch command.
+   * @returns {Response}
+   */
+  convertToFetchResponse() {
```
Comment (axbarsan) [inline src/model/clients/GenericResponse.js:111]: This converts our `GenericResponse` to a `Response` object, which is used by the global `fetch` method.

File: src/model/services/controlplane/appcatalogs/appcatalogs.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  const catalogs = await appsAPI.listApplicationGiantswarmIoV1alpha1AppCatalog(
+    {}
+  );
+  const result = catalogs.items.map(convertAppCatalogCRToCatalog);
+
+  return result;
+}
+
+function convertAppCatalogCRToCatalog(
+  catalog: import('giantswarm-cp-client').ComGithubGiantswarmApiextensionsPkgApisApplicationV1alpha1AppCatalog
```
Comment (axbarsan) [inline src/model/services/controlplane/appcatalogs/appcatalogs.ts:32]: 😬

File: src/model/services/controlplane/appcatalogs/appcatalogs.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+  const catalogs = await appsAPI.listApplicationGiantswarmIoV1alpha1AppCatalog(
+    {}
+  );
+  const result = catalogs.items.map(convertAppCatalogCRToCatalog);
+
+  return result;
+}
+
+function convertAppCatalogCRToCatalog(
```
Comment (axbarsan) [inline src/model/services/controlplane/appcatalogs/appcatalogs.ts:31]: This maps the App catalog CR type to the type that we use internally, so we can have control over what we store in memory.

File: src/model/services/controlplane/appcatalogs/types.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  title: string;
+  description: string;
+  logoURL: string;
+  storage: IStorage;
+}
+
+export interface IStorage {
+  type: string;
+  URL: string;
+}
```
Comment (axbarsan) [inline src/model/services/controlplane/appcatalogs/types.ts:21]: Custom App catalog types, same as the ones on the current GS API.

File: src/model/services/controlplane/appcatalogs/appcatalogs.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import { getBaseConfiguration } from 'model/services/controlplane/base';
+
+/**
+ * Get the all the app catalogs in the installation.
+ * @param client - The HTTP client.
+ */
+export async function getAppCatalogs(
+  client: HttpClient
+): Promise<IAppCatalog[]> {
+  const { ApplicationGiantswarmIoV1alpha1Api } = await import(
```
[the remark this answers is not in the document]
Comment (oponder) [reply src/model/services/controlplane/appcatalogs/appcatalogs.ts:17]: Does that mean we're doing code splitting? Is that just automatically enabled?

[the remark this answers is not in the document]
Comment (axbarsan) [reply src/model/services/controlplane/appcatalogs/appcatalogs.ts:17]: Good call, I was under the impression that we had it enabled. I enabled it now 👍

**Review by oponder — APPROVED** [review]
LGTM! it kind feels like a lot... but I think cool that we're pushing the CP API access in Happa story further.

**Pull request merged**